core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -2525,42 +2522,10 @@ class KafkaApis(val requestChannel: RequestChannel,
         new AlterConfigsRequest(remaining, request.header.apiVersion()),
         response => sendResponse(response.map(_.data())))
     } else {
-      sendResponse(Some(processLegacyAlterConfigsRequest(request, remaining)))
+      throw KafkaApis.shouldAlwaysForward(request)
```

Comment:
How could this happen in kraft mode? Do we need to change some code to make it impossible or is it already impossible? I'd rather do this in a generic way versus having each handler having to handle it.

Reply:
`metadataSupport.canForward()` is always true in kraft and request does not have `envelope` in kraft broker. Hence, it would be better to simplify the code by removing the unnecessary path. for example:
```java
    if (remaining.resources().isEmpty) {
      sendResponse(Some(new AlterConfigsResponseData()))
    } else {
      metadataSupport.forwardingManager.get.forwardRequest(request,
        new AlterConfigsRequest(remaining, request.header.apiVersion()),
        response => sendResponse(response.map(_.data())))
    }
```
BTW, I file https://issues.apache.org/jira/browse/KAFKA-18472 to cleanup `metadataSupport`